### DISCUSSION TITLE
exception in create_node of CloudEndpoint

### DISCUSSION THREAD

**Discussion by zhangxiaomei** [DIRACGrid/DIRAC#6140] [Q&A] [answered]
I am using DIRACv7.3.20. In the CloudDirector log,  when creat node for openstack cloud,  go the error:
"Failed submission to queue cloud-areapd.pd.infn.it_Ce  Invalid argument <email.mime.multipart.MIMEMultipart object at 0x7f9afa7a5c40> for b()"
The problem seems in setting userdata with MIMEMultipart() .  I gave an empty userdata and it is working. The same user data works in v7r1p30.
 Is it caused by different version of libcloud? How to deal with it?

**Comment by sfayer** [comment, marked as answer]
Hi,

Hmm, it does sound like something isn't liking the MIME object... Are there any more details further up in the log, such as the full stack trace? Is this a python3 or python2 based install?

Could you please try converting the userData to a string and see if that makes any difference with the following patch?

https://github.com/DIRACGrid/DIRAC/blob/integration/src/DIRAC/Resources/Cloud/Utilities.py#L36
```patch
-     return S_OK(userData)
+     return S_OK(userData.as_string())
```

As an aside, we're slowly deprecating the old CloudDirector stuff, could you please try out the new CloudComputingElement (which also uses libcloud, but submits with the SiteDirector like any other normal CE) when you have time and let us know if you find anything that doesn't work for your use case?
https://dirac.readthedocs.io/en/rel-v7r3/CodeDocumentation/Resources/Computing/CloudComputingElement.html

Regards,
Simon

**Reply by zhangxiaomei** [reply]
It is python3 based install. There are no more details after, no stack trace in the log:
2022-06-07 07:59:59 UTC WorkloadManagement/CloudDirectorJUNO/CloudEndpoint [140303865972544] VERBOSE: name: DIRAC_EFE8DEDA
 5727 2022-06-07 07:59:59 UTC WorkloadManagement/CloudDirectorJUNO/CloudEndpoint [140303865972544] ERROR: Exception in driver.create_node Invalid
 5728 2022-06-07 07:59:59 UTC WorkloadManagement/CloudDirectorJUNO [140303865972544] ERROR: Failed submission to queue cloud-areapd.pd.infn.it_Ce
 5729  Invalid argument <email.mime.multipart.MIMEMultipart object at 0x7f9afa7a5c40> for b()
 5730 2022-06-07 07:59:59 UTC WorkloadManagement/CloudDirectorJUNO/DIRAC.Core.Tornado.Client.ClientSelector [140303865972544] DEBUG: Trying to au

**Reply by zhangxiaomei** [reply]
With the following changes, it is working:
      - mimeText = MIMEText(contents, mtype, sys.getdefaultencoding())   # default is "utf-8"
      + mimeText = MIMEText(contents, mtype, "ascii")
      -     return S_OK(userData)
      +     return S_OK(userData.as_string())

**Reply by sfayer** [reply]
That looks like a reasonable patch to me, can you please create a pull request against rel-v7r3 with those changes? (or if not, let me know and I'll make a pull request for it).

Regards,
Simon

**Reply by zhangxiaomei** [reply]
I have created the pull request: https://github.com/DIRACGrid/DIRAC/pull/6145
For CloudCE, I will try later after fixing my user data scripts for python3.  Anyway, thank you for reminding me that.
